### PR TITLE
feat!: Switch from Kong proxy to NGINIX proxy

### PR DESCRIPTION
**Pull request by bnevis-i**
BREAKING CHANGE: EdgeX 3.0 replaces the Kong reverse proxy with a generic NGINX reverse proxy.  While the usage model remains the same, there is a minor change to the method by which a JWT authentication token is obtained.

This commit removes Kong and Postgres and replaces with a much lighter weight NGINX and a new security-proxy-auth service.

## PR Checklist
Please check if your PR fulfills the following requirements:

- [ ] I am not introducing a breaking change (if you are, flag in conventional commit message with `BREAKING CHANGE:` describing the break)
- [X] I have fully tested (add details below) this the new feature or bug fix (if not, why?)
- [X] I have opened a PR for the related docs change (if not, why?)
<https://github.com/edgexfoundry/edgex-docs/pull/981>

## Testing Instructions
From edgex-compose, compose-builder directory:

```shell
$ make pull ds-virtual
$ make run ds-virtual
$ watch docker ps -a
```

Ensure that all the services are stable (stay up for 2+ minutes) then try edgex-ui:

```shell
$ make get-token
```

Then browse to http://localhost:4000, paste in the token, and go to Data Center and watch for events (this will ensure that API gateway authentication is working properly.)

### CODE REVIEW

File: compose-builder/add-security.yml [hunk 1]
```diff
@@ -154,87 +153,45 @@ services:
...
+    image: nginx:alpine-slim
+    container_name: edgex-nginx
+    hostname: edgex-nginx
     read_only: true
     restart: always
     networks:
-      - edgex-network
+      edgex-network:
+        aliases:
+          - edgex-kong  # temp patch for edgex-ui-go
```
Comment (hahattan) [inline compose-builder/add-security.yml:166]:
```suggestion
          - edgex-nginx  # temp patch for edgex-ui-go
```

File: compose-builder/add-security.yml [hunk 2]
```diff
@@ -257,12 +214,40 @@ services:
...
       - /tmp/edgex/secrets/security-proxy-setup:/tmp/edgex/secrets/security-proxy-setup:ro,z
       - consul-acl-token:/tmp/edgex/secrets/consul-acl-token:ro,z
     depends_on:
       - security-bootstrapper
       - secretstore-setup
-      - kong
+    security_opt: 
+      - no-new-privileges:true
+
+  authproxy:
```
Comment (hahattan) [inline compose-builder/add-security.yml:228]: how about `proxyauth` or `proxy-auth` which is align with your microservice(binary) name ?

**Review by hahattan — CHANGES_REQUESTED** [review]

File: compose-builder/add-security.yml (hunk 1 above)
Comment (bnevis-i) [reply compose-builder/add-security.yml:166]: Rejected.  This is a workaround for https://github.com/edgexfoundry/edgex-ui-go/pull/592 not being merged yet.

File: compose-builder/add-security.yml (hunk 2 above)
Comment (bnevis-i) [reply compose-builder/add-security.yml:228]: Changed to proxy-auth.

File: compose-builder/add-security.yml (hunk 1 above)
Comment (lenny-intel) [reply compose-builder/add-security.yml:166]: Can we have a TODO comment added to fix this once the UI is fixed?

File: compose-builder/add-security.yml [hunk 3]
```diff
@@ -154,87 +153,45 @@ services:
...
-      - /run
-      - /tmp
     volumes:
       - edgex-init:/edgex-init:ro,z
-      - postgres-config:/tmp/postgres-config:z
-      - kong:/usr/local/kong:z
-      - /tmp/edgex/secrets/security-proxy-setup:/tmp/edgex/secrets/security-proxy-setup:ro,z
+      - nginx-templates:/etc/nginx/templates
+      - nginx-tls:/etc/ssl/nginx
+      #- nginx-conf:/etc/nginx/conf.d
```
Comment (lenny-intel) [inline compose-builder/add-security.yml:181]: Why is this here commented out?

File: compose-builder/add-security.yml [hunk 4]
```diff
@@ -154,87 +153,45 @@ services:
...
-    security_opt: 
-      - no-new-privileges:true
 
-  kong:
-    image: kong:${KONG_VERSION}
-    user: "kong:nogroup" # Note Kong already running as kong user but doesn't define a group
-    container_name: edgex-kong
-    hostname: edgex-kong
+  nginx:
+    image: nginx:alpine-slim
```
Comment (lenny-intel) [inline compose-builder/add-security.yml:158]: Please put the image version in the .env file as NGINX_VERSION and actually pin it to a minor version.

**Review by lenny-intel — COMMENTED** [review]
Why is compose-builder/get-consul-acl-token.sh now empty?

**Comment by lenny-intel** [comment]
Also, I assume the ADD_PROXY_ROUTE setting in add-service-secure-template.yml is still valid place for this?

**Comment by bnevis-i** [comment]
> Why is compose-builder/get-consul-acl-token.sh now empty?

It isn't.  I just chmod'ed it to be executable.

**Comment by lenny-intel** [comment]
> It isn't. I just chmod'ed it to be executable.

Ok. Strange way for the PR to show that...

File: compose-builder/add-security.yml (hunk 3 above)
Comment (bnevis-i) [reply compose-builder/add-security.yml:181]: TODO added.  I made conf.d a tmpfs mount instead.  Removed nginx-conf volume and this commented line.

File: compose-builder/add-security.yml (hunk 4 above)
Comment (bnevis-i) [reply compose-builder/add-security.yml]: Done

**Comment by bnevis-i** [comment]
> Also, I assume the ADD_PROXY_ROUTE setting in add-service-secure-template.yml is still valid place for this?

Should continue to work, but haven't put a lot of effort into testing it.  Consumed by same container and didn't change any of the related scripts.

**Comment by bnevis-i** [comment]
@lenny-intel Changes in https://github.com/edgexfoundry/edgex-compose/pull/345/commits/e249f2e60dedb6f777151b7c72b38e77cb480dd8

File: compose-builder/.env [hunk 5]
```diff
@@ -49,6 +49,7 @@ KONG_VERSION=2.8
 KUIPER_VERSION=1.7-alpine
 MOSQUITTO_VERSION=2.0
 NATS_VERSION=2.9-alpine
+NGINX_VERSION=1.23-alpine-slim
```
Comment (lenny-intel) [inline compose-builder/.env:52]: Need to remove KONG_VERSION and POSTGRES_VERSION

**Review by lenny-intel — CHANGES_REQUESTED** [review]
Minor cleanup left

File: compose-builder/.env (hunk 5 above)
Comment (bnevis-i) [reply compose-builder/.env:49]: Fixed in https://github.com/edgexfoundry/edgex-compose/pull/345/commits/252fc9b8452f2bec2d038c013fc44d351338f6cc

**Review by lenny-intel — APPROVED** [review]
LGTM
